fish-rust/src/abbrs.rs

```diff
@@ -0,0 +1,479 @@
+    pub name: WString,
+
+    /// The key (recognized token) - either a literal or a regex pattern.
+    pub key: WString,
+
+    /// If set, use this regex to recognize tokens.
+    /// If unset, the key is to be interpreted literally.
+    /// Note that the fish interface enforces that regexes match the entire token;
+    /// we accomplish this by surrounding the regex in ^ and $.
+    pub regex: Option<UniquePtr<regex_t>>,
+
+    /// Replacement string.
+    pub replacement: WString,
+
+    /// If set, the replacement is a function name.
+    pub replacement_is_function: bool,
+
+    /// Expansion position.
+    pub position: Position,
+
+    /// If set, then move the cursor to the first instance of this string in the expansion.
+    pub set_cursor_marker: Option<WString>,
+
+    /// Mark if we came from a universal variable.
+    pub from_universal: bool,
+}
+
+impl Abbreviation {
+    // Construct from a name, a key which matches a token, a replacement token, a position, and
+    // whether we are derived from a universal variable.
+    pub fn new(
+        name: WString,
+        key: WString,
+        replacement: WString,
+        position: Position,
+        from_universal: bool,
+    ) -> Self {
+        Self {
+            name,
+            key,
+            regex: None,
+            replacement,
+            replacement_is_function: false,
+            position,
+            set_cursor_marker: None,
+            from_universal,
+        }
+    }
+
+    // \return true if this is a regex abbreviation.
+    pub fn is_regex(&self) -> bool {
+        self.regex.is_some()
+    }
+
+    // \return true if we match a token at a given position.
+    pub fn matches(&self, token: &wstr, position: Position) -> bool {
+        if !self.matches_position(position) {
+            return false;
+        }
+        self.regex
+            .as_ref()
+            .map(|r| r.matches_ffi(&token.to_ffi()))
+            .unwrap_or(self.key == token)
+    }
+
+    // \return if we expand in a given position.
+    fn matches_position(&self, position: Position) -> bool {
+        return self.position == Position::Anywhere || self.position == position;
+    }
+}
+
+/// The result of an abbreviation expansion.
+pub struct Replacer {
+    /// The string to use to replace the incoming token, either literal or as a function name.
+    replacement: WString,
+
+    /// If true, treat 'replacement' as the name of a function.
+    is_function: bool,
+
+    /// If set, the cursor should be moved to the first instance of this string in the expansion.
+    set_cursor_marker: Option<WString>,
+}
+
+impl From<Replacer> for abbrs_replacer_t {
+    fn from(value: Replacer) -> Self {
+        let has_cursor_marker = value.set_cursor_marker.is_some();
+        Self {
+            replacement: value.replacement.to_ffi(),
+            is_function: value.is_function,
+            set_cursor_marker: value.set_cursor_marker.unwrap_or_default().to_ffi(),
```

Comment:
The default value is not used if the cursor is not present, using null failed in `asan` checks